tsl/src/remote/row_by_row_fetcher.c

```diff
@@ -149,73 +262,128 @@ row_by_row_fetcher_complete(RowByRowFetcher *fetcher)
-				 * PQclear() */
-				pfree(response);
-				response = NULL;
-				remote_result_elog(res, ERROR);
+				/*
+				 * Error. The docs say: consult PQerrorMessage() for the reason.
+				 * remote_connection_get_error() will do this for us.
+				 */
+				TSConnectionError err;
+				remote_connection_get_error(fetcher->state.conn, &err);
+				remote_connection_error_elog(&err, ERROR);
 			}
 
-			if (PQresultStatus(res) == PGRES_TUPLES_OK)
+			copy_data.maxlen = copy_data.len;
+			Assert(copy_data.cursor == 0);
+
+			if (fetcher->state.batch_count == 0 && row == 0)
 			{
-				/* fetched all the data */
-				Assert(PQntuples(res) == 0);
+				copy_data_check_header(&copy_data);
+			}
 
+			const AttConvInMetadata *attconv = tuplefactory_get_attconv(fetcher->state.tf);
+			Assert(attconv->binary);
+			const int16 natts = copy_data_read_int16(&copy_data);
+			if (natts == -1)
+			{
 				fetcher->state.eof = true;
-				async_response_result_close(response);
-				response = NULL;
 				break;
 			}
 
-			Assert(PQresultStatus(res) == PGRES_SINGLE_TUPLE);
-			/* Allow creating tuples in alternative memory context if user has set
-			 * it explicitly, otherwise same as batch_mctx */
-			MemoryContextSwitchTo(fetcher->state.tuple_mctx);
-
-			PG_USED_FOR_ASSERTS_ONLY ItemPointer ctid =
-				tuplefactory_make_virtual_tuple(fetcher->state.tf,
-												res,
-												0,
-												PQbinaryTuples(res),
-												&fetcher->batch_values[i * nattrs],
-												&fetcher->batch_nulls[i * nattrs]);
-
 			/*
-			 * This fetcher uses virtual tuples that can't hold ctid, so if we're
-			 * receiving a ctid here, we're doing something wrong.
+			 * There is also one case where no tupdesc attributes are retrieved.
+			 * This is when we do `select count(*) from t`, and
+			 * `enable_partitionwise_aggregate` is 0, so the data node queries
+			 * become `select null from ...` and we should get 1 NULL attribute
+			 * from COPY.
 			 */
-			Assert(ctid == NULL);
+			int16 expected_natts = Max(1, retrieved_natts);
+			if (natts != expected_natts)
+			{
+				elog(ERROR,
+					 "wrong number of attributes for a COPY tuple: expected %d, got %d",
+					 expected_natts,
+					 natts);
+			}
 
-			async_response_result_close(response);
-			response = NULL;
+			Datum *values = &fetcher->batch_values[tupdesc_natts * row];
+			bool *nulls = &fetcher->batch_nulls[tupdesc_natts * row];
+			for (int i = 0; i < tupdesc_natts; i++)
+			{
+				nulls[i] = true;
+			}
+
+			MemoryContextSwitchTo(fetcher->state.tuple_mctx);
+			for (int i = 0; i < retrieved_natts; i++)
+			{
+				const int att = list_nth_int(retrieved_attrs, i) - 1;
+				Assert(att >= 0);
+				Assert(att < tupdesc_natts);
+				const int32 att_bytes = copy_data_read_int32(&copy_data);
+				if (att_bytes == -1)
+				{
+					// NULL
+					values[att] = ReceiveFunctionCall(&attconv->conv_funcs[att],
+													  NULL,
+													  attconv->ioparams[att],
+													  attconv->typmods[att]);
```

Comment:
Turns out we should only call it if the function is not strict. I fixed it and added a citation from postgres docs about this.

Reply:
Sorry, I mean that the result of the function call is NULL (or treated as NULL at least), as evident by the next line `nulls[att] = true` so why do we need to call the function and write the result to `values[att]` in that case?

`STRICT`ness has to do with whether one of the function arguments is NULL or not. Not the result.

Reply:
As I understand it, the Postgres docs tell that non-strict function should be called for NULL inputs:

> Usually, a receive function should be declared STRICT; if it is not, it will be called with a NULL first parameter when reading a NULL input value. The function must still return NULL in this case, unless it raises an error. (This case is mainly meant to support domain receive functions, which might need to reject NULL inputs.)

It must return NULL, but it can also throw an error, that's why we should call it.

Quoting from here: https://www.postgresql.org/docs/current/sql-createtype.html